lib/sinon/test.js

```diff
@@ -55,6 +55,15 @@
                     sandbox.verifyAndRestore();
                 }
 
+                if (result && typeof result.then === "function" && typeof result.catch === "function") {
```

Comment:
> If you simplify this, it will work with more Promise libraries like jQuery (before v3). If you only use .then(onResolve, onReject). I do realize that old jQuery was not Promises/A+ compliant. It's just a thought and almost functionally the same.

+1

>  It also feels like this should be an else-if of the previous if statement. Looking at the code now, a synchronous test that returns a promise will have restore called multiple times.

Speaking strictly with regards to mocha, a test that returns a promise is regarded as asynchronous. I was of the impression that `sandbox.restore` was idempotent but sure, I can move the code around to ensure that doesn't happen, and add a test for it as you suggested.

> If you move your if statement before the typeof OldDone check, you can actually make it so returning a promise is all that is necessary to make tests async instead of return a promise and taking a callback.

I'm not sure I follow. If the user supplied a "done" callback and also returned a Promise, how will that case be handled then?